app/api/v1/helpers_v1.rb

```diff
@@ -20,16 +20,26 @@ def presenter(klass, depth)
 
   def find_object(klass, id)
     klass = klass.classify
-
-    error!("#{klass} not found", 404) unless @object = klass.constantize.where(id: id).first
+    # switched from .find(id) to .where() so that the error!() method will run and, as a result, the content type will be set to JSON
+    error!("#{klass} not found for id=#{id}", 404) unless @object = klass.constantize.where(id: id).first
   end
 
-  def find_objects(klass, ids)
+  def find_objects(klass, params)
     klass = klass.classify
 
-    if ids.any?
-      @objects = klass.constantize.where(id: ids)
-    else
+    if params[:ids].any?
+      @objects = klass.constantize.where(id: params[:ids])
+    elsif params[:query] && params[:query].length > 0
+      # identify invalid parameters (not found in the object)
+      invalid_query_parameters = params[:query].select {|key, value| !klass.constantize.column_names.include? key }
+      if invalid_query_parameters && invalid_query_parameters.length > 0
+        error!("#{klass} query #{params[:query]} has the following invalid parameters: #{invalid_query_parameters.keys}", 400)
+      elsif params[:limit] == 1 # return only one object, the first that meets the query criteria
+        error!("#{klass} not found for query #{params[:query]}", 404) unless @object = klass.constantize.where(params[:query]).first
+      else # return all objects that meet the query criteria
+        @objects = klass.constantize.where(params[:query]).limit(params[:limit]) # a nil limit is ignored by ActiveRecord
```

Comment:
we're protected from SQL injection because Query must be of type Hash:
  params :custom_query do
    optional :limit, type: Integer
#    optional :order, type: String   
    optional :query, type: Hash
  end
end

And, there are tests in place (e.g., context 'query as type String is not valid, must be a Hash' do)